### PR TITLE
Workspace switcher: Expensify is no longer bolded and missing RBR after refreshing the page

### PR DESCRIPTION
**Pull request by ikevin127**
### Details

Because we used to memoize the `brickRoadsForPolicies` and `unreadStatusesForPolicies` [here](https://github.com/Expensify/App/blob/ad9b5d7a2ee54ad6379538b84c322fc29b2a6adb/src/pages/WorkspaceSwitcherPage.tsx#L66-L67) without any variables in the dependency array, the functions using `Onyx.connect` outside of the component lifecycle caused the memoized functions to always return the data from when `allReports` is not populated yet, hence `brickRoadsForPolicies` and `unreadStatusesForPolicies` don't contain any data -> causing:
- the selected **Expensify** to not longer have the bold text style
- RBRs disappear from the other workspaces

Given that both `getWorkspacesBrickRoads` and `getWorkspacesUnreadStatuses` WorkspacesSettingsUtils functions are only used within the WorkspaceSwitcherPage, we can make it such that we actually pass the `reports` and `policies` as arguments directly from the WorkspaceSwitcherPage component, keeping the memoization while leveraging `useOnyx`.

The additional `reportActions` optional argument passed down to `getWorkspacesBrickRoads` and the changes within `WorkspacesSettingsUtils` related to this were done because on Android / iOS **mWeb** the RBR / GBR wouldn't show up when refreshing the page because of [this](https://github.com/Expensify/App/blob/e74c130a20e2cf6a389ed484e60db3e21e22404b/src/libs/WorkspacesSettingsUtils.ts#L55-L66) `Onyx.connect` logic.

**Small refactor bonus:** as a clean-up after doing this we removed the `withOnyx` HOC and all the associated logic and types from `WorkspaceSwitcherPage` as `useOnyx` handles all that 😎

Additionally, the `hasUnreadData` bold style applied to the Workspace switcher workspaces that have unread data was fixed based on https://github.com/Expensify/App/issues/39928#issuecomment-2049311598.

### Fixed Issues

$ https://github.com/Expensify/App/issues/39928
PROPOSAL: https://github.com/Expensify/App/issues/39928#issuecomment-2045740587

### Tests

**Precondition:** There are a few read / unread workspaces, some with RBR (can be invoked by creating scan request with invalid receipt).

1. Open the workspace switcher.
2. Select Expensify in the list (if not selected).
3. Refresh the page.

- [x] Verify that no errors appear in the JS console

### Offline tests

**TLDR:** same as **Tests**.

**Precondition:** There are a few read / unread workspaces, some with RBR (can be invoked by creating scan request with invalid receipt).

1. Open the workspace switcher.
2. Select Expensify in the list (if not selected).
3. Refresh the page.

### QA Steps

**TLDR:** same as **Tests**.

**Precondition:** There are a few read / unread workspaces, some with RBR (can be invoked by creating scan request with invalid receipt).

1. Open the workspace switcher.
2. Select Expensify in the list (if not selected).
3. Refresh the page.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/56457735/777076bb-97de-4317-b960-6f87f0576b84

</details>

**Note:** On Android:mWeb the not found page appearing after refresh happens because of `ngrok` tunneling.

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/56457735/b95f7fd2-df6b-4b99-80cb-8253fd567df6

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/56457735/e6c6c9ad-b960-4ea0-8a05-63255361ca19

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/56457735/70341234-7d14-4298-9fdf-34abaae356ac

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/56457735/bb6307f2-e2b2-4e0c-bd0f-147a2fa4e11a

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/56457735/23194ce9-6c5c-4f53-a75b-fa0b96d3a0cc

</details>

### CODE REVIEW

File: src/libs/WorkspacesSettingsUtils.ts [hunk 1]
```diff
@@ -175,12 +176,22 @@ function getWorkspacesBrickRoads(): Record<string, BrickRoad> {
         }
     });
 
-    Object.values(allReports).forEach((report) => {
+    const actionsByReport: OnyxCollection<ReportActions> = {};
+    Object.keys(reportActions ?? {}).forEach((key) => {
+        if (!reportActions) {
```
Comment (eh2077) [inline src/libs/WorkspacesSettingsUtils.ts:181]: We should check falsy `key` here right?

Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:181]: @eh2077 I guess since report actions can be either `undefined` or `{}` (empty object), would the below condition be ok ?

```ts
if (!reportActions || Object.keys(reportActions ?? {}).length === 0) {
    return;
}
```

Comment (eh2077) [reply src/libs/WorkspacesSettingsUtils.ts:181]: @ikevin127 Sorry, what I want to point out is that I think we don't need to check 
```js
if (!reportActions) {
    return;
}
``` 
here because we already use nullish coalescing operator `Object.keys(reportActions ?? {})`. What do you think?

Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:181]: @eh2077 Oh, I see - cool, I'll remove the check and resolve the other comments in about 30 minutes.

If that's all, you can start working on the checklist and hopefully we can get this ready for final review today.
Thank you!

File: src/libs/WorkspacesSettingsUtils.ts [hunk 2]
```diff
@@ -69,8 +70,8 @@ Onyx.connect({
  * @param report
  * @returns BrickRoad for the policy passed as a param
  */
-const getBrickRoadForPolicy = (report: Report): BrickRoad => {
-    const reportActions = reportActionsByReport?.[report.reportID] ?? {};
+const getBrickRoadForPolicy = (report: Report, actionsByReport?: OnyxCollection<ReportActions>): BrickRoad => {
```
[the remark this answers is not in the document]
Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:74]: All good ?

File: src/libs/WorkspacesSettingsUtils.ts [hunk 3]
```diff
@@ -157,14 +158,14 @@ function checkIfWorkspaceSettingsTabHasRBR(policyID?: string) {
 /**
  * @returns a map where the keys are policyIDs and the values are BrickRoads for each policy
  */
-function getWorkspacesBrickRoads(): Record<string, BrickRoad> {
-    if (!allReports) {
+function getWorkspacesBrickRoads(reports: OnyxCollection<Report>, policies: OnyxCollection<Policy>, reportActions?: OnyxCollection<ReportActions>): Record<string, BrickRoad> {
```
[the remark this answers is not in the document]
Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:165]: All good ?

File: src/libs/WorkspacesSettingsUtils.ts [hunk 4]
```diff
@@ -195,14 +206,14 @@ function getWorkspacesBrickRoads(): Record<string, BrickRoad> {
 /**
  * @returns a map where the keys are policyIDs and the values are truthy booleans if policy has unread content
  */
-function getWorkspacesUnreadStatuses(): Record<string, boolean> {
-    if (!allReports) {
+function getWorkspacesUnreadStatuses(reports: OnyxCollection<Report>): Record<string, boolean> {
```
[the remark this answers is not in the document]
Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:210]: All good ?

File: src/libs/WorkspacesSettingsUtils.ts [hunk 5]
```diff
@@ -175,12 +180,18 @@ function getWorkspacesBrickRoads(): Record<string, BrickRoad> {
         }
     });
 
-    Object.values(allReports).forEach((report) => {
+    const actionsByReport: OnyxCollection<ReportActions> = {};
+    Object.keys(reportActions ?? {}).forEach((key) => {
+        const reportID = CollectionUtils.extractCollectionItemID(key as typeof ONYXKEYS.COLLECTION.REPORT_ACTIONS);
+        actionsByReport[reportID] = reportActions?.[key] ?? {};
```
Comment (ikevin127) [inline src/libs/WorkspacesSettingsUtils.ts:186]: @eh2077 I remember I added that if (!reportActions) early return because otherwise `reportActions.[key]` was throwing possibly `undefined` type error.

I removed the check and instead added optional chaining and fallback to empty object like so: `reportActions?.[key] ?? {}`. Let me know if it's good like this!

Comment (eh2077) [reply src/libs/WorkspacesSettingsUtils.ts:186]: @ikevin127 I'm confused here. Let's say `reportActions` is undefined, it shouldn't enter the foreach loop right?

Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:186]: > Let's say reportActions is undefined, it shouldn't enter the foreach loop right?

@eh2077 Nope, since we fallback to empty object `Object.keys(reportActions ?? {}).forEach...` it enters the foreach loop.

Then: `actionsByReport[reportID]` is expecting type `Record<string, ReportActions | null>` because it's defined above as:
```ts
const actionsByReport: OnyxCollection<ReportActions> = {};
```

If I remove optional chaining from `reportActions?.[key]` I get the following type error:

Screenshot 2024-04-15 at 18 59 44

If I keep optional chaining but don't have the empty object fallback I get the following type error from `actionsByReport[reportID]`:

Screenshot 2024-04-15 at 19 00 31

So we can either keep the if (!reportActions) early return or the optional chaining + fallback to empty object.
If you see other ways to do this please let me know!

Comment (eh2077) [reply src/libs/WorkspacesSettingsUtils.ts:186]: Oh, I see! I like `reportActions?.[key]`. Thank you for you patience.

**Comment by eh2077** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



https://github.com/Expensify/App/assets/117511920/54d6ca06-523c-4c51-878a-78845d6b5e68



</details>

<details>
<summary>Android: mWeb Chrome</summary>



https://github.com/Expensify/App/assets/117511920/30a6d82d-a0a3-4130-911a-5ad61b3ff0a6




</details>

<details>
<summary>iOS: Native</summary>


https://github.com/Expensify/App/assets/117511920/02778b7f-f6ce-4360-bfaf-478360aa436c


</details>

<details>
<summary>iOS: mWeb Safari</summary>



https://github.com/Expensify/App/assets/117511920/05c1ad3a-0e91-4393-bc78-58b676959633




</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



https://github.com/Expensify/App/assets/117511920/61f3ac4c-2f8a-4bbb-be0d-f983c71199fb




</details>

<details>
<summary>MacOS: Desktop</summary>



https://github.com/Expensify/App/assets/117511920/68d2794f-699a-4e0f-bd18-e78281bcafde

</details>

**Review by eh2077 — APPROVED** [review]
Looks good and tested well!

File: src/libs/WorkspacesSettingsUtils.ts [hunk 6]
```diff
@@ -175,12 +180,18 @@ function getWorkspacesBrickRoads(): Record<string, BrickRoad> {
         }
     });
 
-    Object.values(allReports).forEach((report) => {
+    const actionsByReport: OnyxCollection<ReportActions> = {};
+    Object.keys(reportActions ?? {}).forEach((key) => {
+        const reportID = CollectionUtils.extractCollectionItemID(key as typeof ONYXKEYS.COLLECTION.REPORT_ACTIONS);
```
Comment (iwiznia) [inline src/libs/WorkspacesSettingsUtils.ts:185]: Why are we extracting the ID from the key if we have it in the value that we already have?

File: src/libs/WorkspacesSettingsUtils.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (iwiznia) [reply src/libs/WorkspacesSettingsUtils.ts:210]: Same here, this file has all reports already and the only caller is passing all reports, so there's no need to pass this param

**Review by iwiznia — CHANGES_REQUESTED** [review]

File: src/libs/WorkspacesSettingsUtils.ts (hunk 6 above)
Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:185]: Because the forEach `key` looks like this: `reportActions_2298475481263683` but in order to create the object needed by the `getBrickRoadForPolicy` function, we need to extract the ID from the `key`, same as we're doing here:

https://github.com/Expensify/App/blob/ab8c6d3f7b4e73673d2bcaa5e487a3ef3c4f7cfd/src/libs/WorkspacesSettingsUtils.ts#L63-L64

Comment (iwiznia) [reply src/libs/WorkspacesSettingsUtils.ts:185]: I know that. I am saying that each entry in `reportActions` has a key with the ID already, so you don't need to call `extractCollectionItemID` on its key to get that data, you can get it directly from the reportAction object.

File: src/libs/WorkspacesSettingsUtils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:165]: > which is already what we did before, so this param is useles

Not quite, you can find more details in the [proposal's](https://github.com/Expensify/App/issues/39928#issuecomment-2045740587) RCA / alternative solution, as to why we're not using the local `allReports` and instead passing it down from the component `reports` coming from the `useOnyx` hook.

**TLDR:** Because local action file's `allReports` is not hooked to the component's lifecycle which causes a delay in showing the workspace switcher data when refreshing the page (see issue and selected proposal). Passing down `reports` from the component via `useOnyx` makes it such that upon refresh the UI is instant (no delays).

[the remark this answers is not in the document]
Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:165]: > same for reportActions, do we need to pass them down or should we subscribe to them here instead

I did this (additionally from my proposal) because on mWeb the RBR/GRB wouldn't show up in workspace switcher upon page refresh based on local action file's `reportActionsByReport`.

File: src/libs/WorkspacesSettingsUtils.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:210]: See [above (comment)](https://github.com/Expensify/App/pull/40197/files#r1567797163), it's for the same reason as detailed there.

**TLDR:** Using local action file's `allReports` would create a delay in the UI since `Onyx.connect` is decoupled from component's lifecycle.

File: src/libs/WorkspacesSettingsUtils.ts (hunk 6 above)
Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:181]: I see what you mean. If you're asking why am I doing it like this: because it works and it's used above in the component in similar fashion.

If you propose another way of writing it, please share it and I'll replace it.

File: src/libs/WorkspacesSettingsUtils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (iwiznia) [reply src/libs/WorkspacesSettingsUtils.ts:161]: Oh I see.... so isn't this same problem going to happen with basically any component using the methods on this file that use `allReports`  and any other collection we subscribe here that also has `waitForCollectionCallback` set to true?

File: src/libs/WorkspacesSettingsUtils.ts (hunk 6 above)
Comment (iwiznia) [reply src/libs/WorkspacesSettingsUtils.ts:181]: Not sure if we use underscore still but `_.pluck(reportActions, 'reportActionID')`

Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:181]: As per [TS - Migration Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/TS_STYLE.md#migration-guidelines):

> Deprecate the usage of `underscore`. **Use vanilla methods from JS instead**. Only use `lodash` when there is no easy vanilla alternative (eg. lodashMerge). eslint: [no-restricted-imports](https://eslint.org/docs/latest/rules/no-restricted-imports)

This is why I went with `Object.keys(reportActions ?? {}).forEach`.

File: src/libs/WorkspacesSettingsUtils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:161]: As of now `getWorkspacesBrickRoads` and `getWorkspacesUnreadStatuses` are only used within the `WorkspaceSwitcherPage` component.

It is true that if these functions were to be used in other components and keep the external `Onyx.connect` data implementation, that could cause similar stale data problems as it happened with the issue this PR fixes.

If you look at the issue's proposals, you'll see the differences in UI between the other proposals and why my alternative solution using `useOnyx` was selected (UI is instant).

I think the changes from this PR will make it such that for the future, if these functions are used in other components it will enforce similar usage by passing `reports`, `policies` and optionally `reportActions` straight from the component (lifecycle) either from component's props (`withOnyx` HOC) or from the recent `useOnyx` hook implementation.

**Comment by ikevin127** [comment]
@iwiznia I'll handle the conflicts since there's been some changes and some of your previous comments don't apply anymore. After I'll push please check the code again and let me know if all good for approval or you require other changes.

File: src/libs/WorkspacesSettingsUtils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (iwiznia) [reply src/libs/WorkspacesSettingsUtils.ts:149]: > I think the changes from this PR will make it such that for the future, if these functions are used in other components it will enforce similar usage by passing reports, policies and optionally reportActions straight from the component

I am confused, since this PR makes the param optional, people can misuse it by not passing the param. So if this bug will show up for anyone not passing the param, we should remove its optionality.

[the remark this answers is not in the document]
Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:149]: Indeed, makes more sense, to have the `getWorkspacesBrickRoads`'s `reportActions` **required** but keep the `getBrickRoadForPolicy`'s `altReportActions` **optional** since `getBrickRoadForPolicy` is also used here:

https://github.com/Expensify/App/blob/26c9eba8c209090b5c17f6ba8c5a6c318555ad05/src/libs/WorkspacesSettingsUtils.ts#L113

in the `getChatTabBrickRoad` function which we don't want to change, keep it backwards compatible.

cc @iwiznia

File: src/libs/WorkspacesSettingsUtils.ts (hunk 6 above)
Comment (iwiznia) [reply src/libs/WorkspacesSettingsUtils.ts]: But is the bug present in that call, given we are not passing the actions?

Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts]: No - if you're referring to the `getBrickRoadForPolicy` function discussed [here](https://github.com/Expensify/App/pull/40197#discussion_r1571181353), this is used only in one component BottomTabBar to display RBR dot above the Chats icon:

Screenshot 2024-04-18 at 21 30 21

Here's a video showing that it works as expected without any changes, even when refreshing the page (our issue with the workspace switcher):

https://github.com/Expensify/App/assets/56457735/abc626b8-2252-490b-a070-862faaa0d93c

cc @iwiznia

File: src/libs/WorkspacesSettingsUtils.ts [hunk 7]
```diff
@@ -54,11 +55,12 @@ Onyx.connect({
...
 /**
- * @param report
- * @returns BrickRoad for the policy passed as a param
+ * @param altReportActions Replaces (local) allReportActions used within (local) function getWorkspacesBrickRoads
+ * @returns BrickRoad for the policy passed as a param and optionally actionsByReport (if passed)
  */
-const getBrickRoadForPolicy = (report: Report): BrickRoad => {
-    const reportActions = allReportActions?.[`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${report.reportID}`] ?? {};
+const getBrickRoadForPolicy = (report: Report, altReportActions?: OnyxCollection<ReportActions>): BrickRoad => {
+    const finalReportActions = isEmptyObject(altReportActions) ? allReportActions : altReportActions;
```
[the remark this answers is not in the document]
Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:62]: @iwiznia Correct, are you suggesting doing this instead ?

`const reportActions = (altReportActions ?? allReportActions)?.['${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${report.reportID}'] ?? {};`

[the remark this answers is not in the document]
Comment (iwiznia) [reply src/libs/WorkspacesSettingsUtils.ts:62]: `altReportActions ?? allReportActions` I assume this is a typo?

[the remark this answers is not in the document]
Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts:62]: `altReportActions` is the alternative reportActions coming from param and `allReportActions` are the local reportActions coming from Onyx.connect.

`altReportActions ?? allReportActions` suggestion was made because you pointed out that `isEmptyObject(altReportActions) ? allReportActions : altReportActions` is not necessary since if `altReportActions` is not passed it means it's undefined / null, therefore `??` operator is enough to fallback to `allReportActions` if the param is not passed.

Let me know if you have a problem with the `altReportActions` naming and what suggestions you'd have if so.
Otherwise let me know what else is required for you to approve this.

[the remark this answers is not in the document]
Comment (iwiznia) [reply src/libs/WorkspacesSettingsUtils.ts:62]: Oh, I am blind, did not notice the `t` there 🤦 

Yeah, that's what I was suggesting

[the remark this answers is not in the document]
Comment (ikevin127) [reply src/libs/WorkspacesSettingsUtils.ts]: @iwiznia Done, removed `isEmptyObject` check and changed it to:

`const reportActions = (altReportActions ?? allReportActions)?.['${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${report.reportID}'] ?? {};`

Works as expected!

**Review by iwiznia — APPROVED** [review]
